src/Workspaces/Remote/ServiceHub/Services/CodeAnalysisService_CodeLens.cs

```diff
@@ -63,12 +87,140 @@ public Task<string> GetFullyQualifiedName(DocumentId documentId, TextSpan textSp
+        }
+
+        /// <summary>
+        /// it tracks relevant changes on workspace for the given document.
+        /// 
+        /// better place for this is in ICodeLensContext but CodeLens OOP doesn't provide a way to call back to codelens OOP from
+        /// VS so, this for now will be in Roslyn OOP
+        /// </summary>
+        private class WorkspaceChangeTracker
+        {
+            private static readonly TimeSpan s_delay = TimeSpan.FromMilliseconds(100);
+
+            private readonly object _gate;
+
+            private readonly CodeAnalysisService _owner;
+            private readonly Workspace _workspace;
+            private readonly DocumentId _documentId;
+
+            private VersionStamp _lastVersion;
+            private ResettableDelay _resettableDelay;
+
+            public static async Task TrackAsync(CodeAnalysisService owner, Workspace workspace, DocumentId documentId, CancellationToken cancellationToken)
+            {
+                var document = workspace.CurrentSolution.GetDocument(documentId);
+                if (document == null)
+                {
+                    return;
+                }
+
+                // if anything under the project this file belong to changes, then invalidate the code lens so that it can refresh
+                var dependentVersion = await document.Project.GetDependentVersionAsync(cancellationToken).ConfigureAwait(false);
+                var _ = new WorkspaceChangeTracker(owner, workspace, documentId, dependentVersion);
+            }
+
+            private WorkspaceChangeTracker(CodeAnalysisService owner, Workspace workspace, DocumentId documentId, VersionStamp dependentVersion)
+            {
+                _gate = new object();
+
+                _owner = owner;
+                _workspace = workspace;
+                _documentId = documentId;
+
+                _lastVersion = dependentVersion;
+                _resettableDelay = ResettableDelay.CompletedDelay;
+
+                ConnectEvents(subscription: true);
+            }
+
+            private void ConnectEvents(bool subscription)
+            {
+                // this is only place lock is used.
+                // we have a lock here so that subscription and unsubscription of the two events
+                // are happening atomic, but that doesn't mean there is no possiblity of race here
+                // theoradically, there can be a race if connection got disconnected before we subscribe
+                // to OnRpcDisconnected but already in the subscription code path.
+                // but there is no easy way to solve the problem unless Rpc itself provide things like
+                // subscribe only if connection still alive or something
+                lock (_gate)
+                {
+                    if (subscription)
+                    {
+                        _owner.Disconnected += OnDisconnected;
+                        _workspace.WorkspaceChanged += OnWorkspaceChanged;
```

Comment:
❓ If we reverse the order of these two subscriptions, is the lock still necessary?

Reply:
actually, there is another race since connection can disconnected once we are subscribing to events. 

I should check whether thing got disconnected while I am subscribing to events.